core/src/main/java/org/apache/iceberg/TableScanContext.java

```diff
@@ -351,14 +352,15 @@ TableScanContext planWith(ExecutorService executor) {
         toSnapshotId,
         executor,
         fromSnapshotInclusive,
-        metricsReporter);
+        metricsReporters);
   }
 
-  MetricsReporter metricsReporter() {
-    return metricsReporter;
+  Collection<MetricsReporter> metricsReporter() {
+    return metricsReporters;
   }
 
   TableScanContext reportWith(MetricsReporter reporter) {
+    metricsReporter().add(reporter);
```

Comment:
If I remember correctly, this class is immutable and we create a new instance on every call. We should probably follow what is done for updating properties in `withOption`, where we copy existing options into a new list before adding.

```
TableScanContext reportWith(MetricsReporter reporter) {
  ImmutableList.Builder<MetricsReporter> builder = ImmutableList.builder();
  builder.addAll(metricsReporters);
  builder.add(reporter);
  List<MetricsReporter> newMetricsReporters = builder.build();
  return new TableScanContext(
      snapshotId,
      rowFilter,
      ignoreResiduals,
      caseSensitive,
      colStats,
      projectedSchema,
      selectedColumns,
      options,
      fromSnapshotId,
      toSnapshotId,
      planExecutor,
      fromSnapshotInclusive,
      newMetricsReporters);
}
```